### ISSUE TITLE
Getting Drupal FunctionalJavascript Unit Test in PHPStorm

### ISSUE THREAD

**Issue by mdolnik-eelzee** [drud/ddev#1579]
I'm not sure if this is the best place to ask, but I am trying to get Drupal FunctionalJavascript Unit Testing working in PHPStorm.

I am running a Drupal 8.6.10 site in PHP 7.2 and set up Docker/PHPUnit in PHPStorm using instructions from the following sources:
- <https://www.youtube.com/watch?v=I7aGWO6K3Ho>
- <https://glamanate.com/blog/running-drupals-phpunit-test-suites-ddev>
- <https://glamanate.com/blog/running-drupals-functionaljavascript-tests-ddev>

The above allowed me to get regular unit tests (such as `ConfigFileReaderTest`) working by pressing the "play button" in the PHPStorm UI, but when I try to run a `FunctionalJavascript` such as `BigPipeRegressionTest` I get the following result/error:
```
[docker://drud/ddev-webserver:v1.5.2/]:php /opt/project/vendor/phpunit/phpunit/phpunit --configuration /opt/project/public_html/core /opt/project/public_html/core/modules/big_pipe/tests/src/FunctionalJavascript --teamcity
PHPUnit 6.5.14 by Sebastian Bergmann and contributors.

Testing Drupal\Tests\big_pipe\FunctionalJavascript\BigPipeRegressionTest

Drupal\Core\Installer\Exception\AlreadyInstalledException : <ul>
<li>To start over, you must empty your existing database and copy <em>default.settings.php</em> over <em>settings.php</em>.</li>
<li>To upgrade an existing installation, proceed to the <a href="/update.php">update script</a>.</li>
<li>View your <a href="http://web">existing site</a>.</li>
```

The odd thing is if I run the same test in the `ddev ssh` and replace the `/opt` path with the ddev `/var` path it works (well 3/4 tests pass):
```
$ php /var/www/html/vendor/phpunit/phpunit/phpunit --configuration /var/www/html/public_html/core /var/www/html/public_html/core/modules/big_pipe/tests/src/FunctionalJavascript
PHPUnit 6.5.14 by Sebastian Bergmann and contributors.

Testing /var/www/html/public_html/core/modules/big_pipe/tests/src/FunctionalJavascript
..F.                                                                4 / 4 (100%)

Time: 2.12 minutes, Memory: 4.00MB

There was 1 failure:

1) Drupal\Tests\big_pipe\FunctionalJavascript\BigPipeRegressionTest::testMessages_2712935
User <em class="placeholder">o9kNlFTJ</em> successfully logged in.
Failed asserting that false is true.

/var/www/html/public_html/core/tests/Drupal/KernelTests/AssertLegacyTrait.php:31
/var/www/html/public_html/core/tests/Drupal/Tests/UiHelperTrait.php:254
/var/www/html/public_html/core/modules/big_pipe/tests/src/FunctionalJavascript/BigPipeRegressionTest.php:154

FAILURES!
Tests: 4, Assertions: 25, Failures: 1.
```

I can't find anything online about the `AlreadyInstalledException` in this scenario except that people find it to be a vague error that can mean a few things and it mostly occurs when communication with the db cannot be made.

Another difference between the tests I noticed is during the tests, a `[drupal root]\sites\simpletest\[some number]` folder is created and:
- The PHP Storm tests only create a `services.yml` file and `settings.php` file where the `settings.php` does not have any db credentials in it. Also the folder remains after test completion.
- The `ddev ssh` tests end up with the two files above, but the `settings.php` file has the db credentials and directories: `files`, `private`, and `temp` are created. The folder gets removed after test completion.

**Comment by rfay** [comment]
@mglaman any suggestions on this?

**Comment by rfay** [comment]
I spent some time with this, got to where you got (which was really, really hard).

I'm pretty sure the issue is the long-running https://www.drupal.org/project/drupal/issues/2459671

Also: https://drupal.stackexchange.com/questions/261947/functional-test-fails-with-phpunit-on-local

But I tried changing SIMPLETEST_DB:
```
    <env name="SIMPLETEST_DB" value="mysql://db:db@db:3306/db"/>
```

and it didn't do the job for me.

**Comment by rfay** [comment]
OK, I think I see what's going on.

PHPStorm is not doing anything that even slightly works with ddev. It's running its own version of the web container. I see this:

```
5f447e338450        ba73c300777d                                    "/usr/bin/php7.2 /va…"   48 seconds ago      Up 47 seconds (unhealthy)   80/tcp, 8025/tcp                                                                           exciting_mcnulty
```

And it's running it without any of the config or state or webserver setup that ddev would apply.

But the real issue is that a random container run like this doesn't share the ddev_default network, so the database is not accessible from inside the (separate) web container that PHPStorm sets up. 

And these names/IPs are not accessible: `ddev-d8composer-db`, `db`, and `172.18.0.4`

There are some PHPStorm/Jetbrains help articles that may provide a way out:

https://intellij-support.jetbrains.com/hc/en-us/community/posts/115000058610-How-to-get-the-new-Docker-CLI-interpreter-to-connect-to-databases-in-other-containers

**Comment by mglaman** [comment]
This has worked fine with local PHPUnit and PhpStorm. I've never executed remote PHPUnit.

**Comment by rfay** [comment]
Basically I don't think this can ever work with ddev without some major juggling. PHPStorm does not (yet?) have a way to configure the docker container to be on any network but the default network, and thus the extra container that PHPStorm is spinning up cannot reach the db container.  And really, I don't think it's going to reach the chromedriver container after that, so this is just the beginning.

I experimented with setting the extra_hosts to an explicit IP, but that can't work because the IP is not reachable. 

<img width="1012" alt="Edit_Docker_Container_Settings_and_Preferences_and_d8composer____workspace_d8composer__-_____web_core_phpunit_xml__d8composer_" src="https://user-images.githubusercontent.com/112444/57589581-0f526600-74e2-11e9-9f5d-fff5e22f3900.png">

I also tried doing a "link", which seemed like it had promise:

<img width="691" alt="Edit_Docker_Container_Settings_and_Preferences_and_d8composer____workspace_d8composer__-_____web_core_phpunit_xml__d8composer_" src="https://user-images.githubusercontent.com/112444/57589613-71ab6680-74e2-11e9-8853-b5eeed1a14d4.png">

But got the obvious error:

<img width="1096" alt="d8composer____workspace_d8composer__-_____web_core_phpunit_xml__d8composer_" src="https://user-images.githubusercontent.com/112444/57589617-7ff98280-74e2-11e9-9265-9e2f0b49fe9d.png">


But oh my goodness just when I decided it was impossible... The "network mode" is actually a mislabeled field that should be "docker network". So you set "network mode" to "ddev_default" and that's the answer. 

<img width="521" alt="Edit_Docker_Container_Settings_and_Preferences" src="https://user-images.githubusercontent.com/112444/57589717-60168e80-74e3-11e9-86cb-520c1a4cbf00.png">


I think you can get there from here, not sure if you want to. We've wandered through:

* Horrible Drupal misleading error message
* MIslabeled obscure "docker network" configuration
* PHPStorm actually doing something really quite different from what you're hoping for or expecting 

But...

<img width="1393" alt="d8composer____workspace_d8composer__-_____web_core_phpunit_xml__d8composer_" src="https://user-images.githubusercontent.com/112444/57589756-ba175400-74e3-11e9-9182-00e5bbc606b6.png">

**Comment by rfay** [comment]
No response from OP, so closing, happy to continue the conversation. But it's clear that what PHPStorm is doing with Docker really isn't related to ddev or useful as far as I can tell. It *can* be made to run an extra container that happens to be ddev-webserver, but that's as close as we come.

**Comment by mdolnik-eelzee** [comment]
Sorry for the late response and thank you for taking your time on helping me.

The instructions above definitely helped, but I was still having some issues.

The help above managed to fix the issue with the `AlreadyInstalledException` error, and the `[drupal root]\sites\simpletest\[some number]` folders have the db credentials and file folders, but now all of the tests I run seem to be failing in an unreliable manner.

Running the `action/tests/src/FunctionalJavascript` that @rfay ran above gave me this result the first time:
![2019-05-16_17h21_21](https://user-images.githubusercontent.com/20405268/57895259-8ef07580-77ff-11e9-9d02-8c1beb6ee06c.png)
Then without changing anything and running the test again I got:
![2019-05-16_17h22_40](https://user-images.githubusercontent.com/20405268/57895267-9ca5fb00-77ff-11e9-9a92-6a4f7cb8c42d.png)
And it seemed to be a roll of the dice which error above I got.

So I tried adding the chromedriver to the Docker Container Settings
![2019-05-16_17h45_33](https://user-images.githubusercontent.com/20405268/57895839-864d6e80-7802-11e9-8229-0601d721c705.png)
...and it passed.... once...

When I ran the test again I got a new error:
![2019-05-16_17h47_31](https://user-images.githubusercontent.com/20405268/57895887-b1d05900-7802-11e9-903c-390242c1585d.png)

Then I ran it again and got the `User successfully logged in Failing asserting that false is true` error again...

Then it passed, then it failed, then it failed, then it passed 6 times in a row, then it failed, ect...

So in a nutshell, now it doesn't fail every time, but it doesn't pass every time, does anyone have any suggestions on what I can try or look into?

Also in your screenshots it showed your paths in the `/var/www/html` format and not the `/opt/project` format, any time I try the `/var/www/html` format the tests cannot find the correct paths, could this be my issue? What would I need to do to set my config to get the `/var/www/html` format working?

**Comment by rfay** [comment]
I mounted the project to /var/www/html, which is what ddev would do, so that's the reason for that. 

Also, because PHPStorm is running the ddev-webserver container without any of ddev's context, I had to set the php interpreter to the one that matched my setup, /usr/bin/php72 if I remember right. 

Remember, PHPStorm is naively using the ddev-webserver container without any context (no nginx setup, no PHP version setup, no nothing). And it's *not* running in ddev's web container, it's running with a completely separate container. 

My screenshot above shows the mappings, it's in the "Edit docker container settings" dialog. 

Again, PHPStorm isn't doing you any favors. I'm a great fan of PHPStorm and have been for years. But what's it's doing here is intended for a trivial web-only container that's doing php only. It's not intended for integration with a full setup like ddev.

**Comment by mdolnik-eelzee** [comment]
> I mounted the project to /var/www/html, which is what ddev would do, so that's the reason for that.

I also have my project mounted to `/var/www/html`...

Dealing with Drupal FunctionalJavascript Unit Tests is the first time I've seen or dealt with the `/opt/project` paths using ddev and I am currently using that because that brings me the closest to getting it working.

Even if I change all phpstorm settings paths to `/var/www/html` it complains about not finding the file path(s) when I run the tests and complains about not finding the phpunit path in `Settings | Languages & Frameworks | PHP | Test Frameworks`

Regardless I don't think this is an issue strictly with phpstorm, as running the test manually in the `ddev ssh` environment yields the same results for me (random chance of passing tests)...

ie: With my current setup, I get the same results with right-clicking on `core/modules/action/tests` and selecting `Run 'tests (PHPUnit)'` which runs the command:
```
Testing started at 10:44 AM ...
[docker://drud/ddev-webserver:v1.5.2/]:php /opt/project/vendor/phpunit/phpunit/phpunit --configuration /opt/project/public_html/core /opt/project/public_html/core/modules/action/tests/src/FunctionalJavascript --teamcity
PHPUnit 6.5.14 by Sebastian Bergmann and contributors.

Testing /opt/project/public_html/core/modules/action/tests/src/FunctionalJavascript
```
... I get the same results when manually calling the same tests manually in the `ddev ssh` environment:
```
php /var/www/html/vendor/phpunit/phpunit/phpunit --configuration /var/www/html/public_html/core /var/www/html/public_html/core/modules/action/tests/src/FunctionalJavascript
PHPUnit 6.5.14 by Sebastian Bergmann and contributors.

Testing /var/www/html/public_html/core/modules/action/tests/src/FunctionalJavascript
```

**Comment by rfay** [comment]
Sorry, what you're doing isn't really related to something that ddev can support; it's obviously something that can be done, but it's not related to ddev, so I'm going to have to let you chase it. I spent a few hours making it work, found it wasn't related to ddev, reported the results. 

I recommend just using @mglaman 's excellent approach in https://glamanate.com/blog/running-drupals-functionaljavascript-tests-ddev and moving on :) But I wish you the best no matter how you do it.

**Comment by mdolnik-eelzee** [comment]
Ok another update, I tried closing ALL other running DDEV projects and...  *mostly* success!

After ensuring the DDEV project I'm testing on is the only one running, I ran the same `core/modules/action/tests/src/FunctionalJavascript` test over and over on both the terminal and PHPStorm UI and had the following results:

- PHPStorm UI tests passed: 25/25
- Terminal tests passed: 22/25

With the failing tests being the following issues:
```
There was 1 error:

1) Drupal\Tests\action\FunctionalJavascript\ActionFormAjaxTest::testActionConfigurationWithAjax Error: Call to a member function setValue() on null

/var/www/html/public_html/core/modules/action/tests/src/FunctionalJavascript/ActionFormAjaxTest.php:48
```
```
There was 1 error:

1) Drupal\Tests\action\FunctionalJavascript\ActionFormAjaxTest::testActionConfigurationWithAjaxBehat\Mink\Exception\ResponseTextException: The text "The action has been successfully saved." was not found anywhere in the text of the current page.

/var/www/html/vendor/behat/mink/src/WebAssert.php:785
/var/www/html/vendor/behat/mink/src/WebAssert.php:262
/var/www/html/public_html/core/modules/action/tests/src/FunctionalJavascript/ActionFormAjaxTest.php:54
```
```
There was 1 failure:

1) Drupal\Tests\action\FunctionalJavascript\ActionFormAjaxTest::testActionConfigurationWithAjaxUser <em class="placeholder">bPJ6SMiK</em> successfully logged in. Failed asserting that false is true.

/var/www/html/public_html/core/tests/Drupal/KernelTests/AssertLegacyTrait.php:31
/var/www/html/public_html/core/tests/Drupal/Tests/UiHelperTrait.php:254
/var/www/html/public_html/core/modules/action/tests/src/FunctionalJavascript/ActionFormAjaxTest.php:27
```

So instead of 90% of my tests failing, 90% are passing, I still don't know why the failed tests failed, but these are margins that I can live with, I guess...

So presumably the issue with the other DDEV projects running had to do with the DDEV Project I was testing having a `docker-compose.chromedriver.yml` (for unit tests) and the other running project(s) having `docker-compose.selenium.yml` (for behat tests) and both are running on port 4444 which I assume is where the conflict was occurring.

So I guess we can consider this issue resolved, thanks everyone for your help.

**Comment by rfay** [comment]
Yay, glad it's working for you!

**Comment by rfay** [comment]
Just a note that I returned to this with current PHPStorm and didn't completely succeed, but did have to add composer autoloader info this time, don't remember having to do that last time:
<img width="1043" alt="Test_Frameworks_and_d8composer____workspace_d8composer__-_____web_core_modules_action_tests_src_FunctionalJavascript_ActionFormAjaxTest_php" src="https://user-images.githubusercontent.com/112444/68596954-5998ad80-0459-11ea-8721-ab2994b932e5.png">
